### PR TITLE
Further optimizing Google Analytics snippet

### PR DESCRIPTION
**Pull request by BLooperZ**
__EDIT__: it started to be a real mess in this post so i cleared it up

this is a shorter (and still safe) version of the google analytics snippet - 322 charcters against 346
it does exactly the same - except adding the script after the first script in the document rather than before it. only one function changed - `insertBefore` to `appendChild` to eliminate the need for one of the parameters, all the rest is done by syntax manipulation __ONLY__
this is a bit more readable version (same code just added whitespace:
```
(function (b, o, i, l) {
    GoogleAnalyticsObject = l;
    (b[l] = b[l] || function () {
        (b[l].q = b[l].q || []).push(arguments)}
    ).l = +new Date;
    o.getElementsByTagName(i)[0].parentNode.appendChild(o.createElement(i))
    .src='//www.google-analytics.com/analytics.js'}
(window, document, 'script', 'ga'));ga('create', 'UA-XXXXX-X', 'auto');ga('send', 'pageview');
```
let go line by line:
__new__ `(function (b, o, i, l) {` __old__ `(function(b,o,i,l,e,r)`
not much to say we need less parameters
__new__ `GoogleAnalyticsObject = l;` __old__ `b.GoogleAnalyticsObject=l;`
the window object is assumed if no object specified (except for conditional statements)
__new__ `(b[l] = b[l] || function () {` __old__ `b[l] || (b[l] = function () {`
shorter syntax and we can use it 2 lines later
__new__ `(b[l].q = b[l].q || []).push(arguments)}` __old__ `(b[l].q = b[l].q || []).push(arguments)}`
no change here
__new__ `).l = +new Date;` __old__ `);b[l].l = +new Date;`
we use the functio0n it self as the object instead of referencing it
__new__ 
```
o.getElementsByTagName(i)[0].parentNode.appendChild(o.createElement(i))
.src='//www.google-analytics.com/analytics.js'}
```
__old__ 
```
e=o.createElement(i);r=o.getElementsByTagName(i)[0];
e.src='//www.google-analytics.com/analytics.js';
r.parentNode.insertBefore(e,r)}
```
by changing to appendChild we do not need reference anymore and we declare anything inside the functions - using the original value and saving the need for a parameter
__new__ `(window, document, 'script', 'ga'));ga('create', 'UA-XXXXX-X', 'auto');ga('send', 'pageview');` __old__ `(window, document, 'script', 'ga'));ga('create','UA-XXXXX-X','auto');ga('send','pageview');`
also the same

I've checked this script with google analytics debugger, also with multiple trackers
it works (no reason why not)
@mathiasbynens - feel free to update your article about it if you like

### CODE REVIEW

File: dist/index.html [hunk 1]
```diff
@@ -29,11 +29,10 @@
 
         <!-- Google Analytics: change UA-XXXXX-X to be your site's ID. -->
         <script>
-            (function(b,o,i,l,e,r){b.GoogleAnalyticsObject=l;b[l]||(b[l]=
+            (function(b,o,i,l){b.GoogleAnalyticsObject=l;b[l]||(b[l]=
             function(){(b[l].q=b[l].q||[]).push(arguments)});b[l].l=+new Date;
-            e=o.createElement(i);r=o.getElementsByTagName(i)[0];
-            e.src='//www.google-analytics.com/analytics.js';
-            r.parentNode.insertBefore(e,r)}(window,document,'script','ga'));
+            o.getElementsByTagName(i)[0].parentNode.appendChild(o.createElement(i)).src='//www.google-analytics.com/analytics.js'}
```
Comment (BLooperZ) [inline dist/index.html]: just all of the three lines above in one line

Comment (arthurvr) [inline dist/index.html]: The lines were wrapped on purpose I think.

Comment (BLooperZ) [inline dist/index.html]: It does exactly the same, just written in one command instead of four

File: dist/index.html [hunk 2]
```diff
@@ -29,12 +29,9 @@
 
         <!-- Google Analytics: change UA-XXXXX-X to be your site's ID. -->
         <script>
-            (function(b,o,i,l,e,r){b.GoogleAnalyticsObject=l;b[l]||(b[l]=
-            function(){(b[l].q=b[l].q||[]).push(arguments)});b[l].l=+new Date;
-            e=o.createElement(i);r=o.getElementsByTagName(i)[0];
-            e.src='//www.google-analytics.com/analytics.js';
-            r.parentNode.insertBefore(e,r)}(window,document,'script','ga'));
-            ga('create','UA-XXXXX-X','auto');ga('send','pageview');
+            (function(b,o,i){b.GoogleAnalyticsObject='g';(b.g||(b.g=function(){(b.g.q=b.g.q||[]).push(arguments)})).l=+new Date;
```
Comment (BLooperZ) [inline dist/index.html]: we can save 2 more characters by wrapping the function making a total of 315

**Comment by BLooperZ** [comment]
here is versions of this optimized script to choose from:
(someone please check the size when gzipping this, i don't know how)

__perfect solution - safe for all - same as in the pull request file__ - 321 characters
```
(function(b,o,i,l){b.GoogleAnalyticsObject=l;(b[l]=b[l]||function(){(b[l].q=b[l].q||[]).push(arguments)}).l=+new Date;
o.getElementsByTagName(i)[0].parentNode.appendChild(o.createElement(i)).src='//www.google-analytics.com/analytics.js'}
(window,document,'script','g'));g('create','UA-XXXXX-X','auto');g('send','pageview');
```

__if you're not going to change the variable name__ - 313 characters
```
(function(b,o,i){b.GoogleAnalyticsObject='g';(b.g=b.g||function(){(b.g.q=b.g.q||[]).push(arguments)}).l=+new Date;
o.getElementsByTagName(i)[0].parentNode.appendChild(o.createElement(i)).src='//www.google-analytics.com/analytics.js'}
(window,document,'script'));g('create','UA-XXXXX-X','auto');g('send','pageview');
```

__if you don't care about Firefox < 9, BlackBerry OS 5 and OmniWeb 622__ - 287 characters
```
(function(b,o,l){b.GoogleAnalyticsObject=l;(b[l]=b[l]||function(){(b[l].q=b[l].q||[]).push(arguments)}).l=+new Date;
o.scripts.appendChild(o.createElement('script')).src='//www.google-analytics.com/analytics.js'}
(window,document,'g'));g('create','UA-XXXXX-X','auto');g('send','pageview');
```

__if you don't care about Firefox < 9, BlackBerry OS 5 and OmniWeb 622 nor defining variable name__ - 279 characters
```
(function(b,o){b.GoogleAnalyticsObject='g';(b.g=b.g||function(){(b.g.q=b.g.q||[]).push(arguments)}).l=+new Date;
o.scripts.appendChild(o.createElement('script')).src='//www.google-analytics.com/analytics.js'}
(window,document));g('create','UA-XXXXX-X','auto');g('send','pageview');
```

__if your page MUST got head__ - 284 characters
```
(function(b,o,l){b.GoogleAnalyticsObject=l;(b[l]=b[l]||function(){(b[l].q=b[l].q||[]).push(arguments)}).l=+new Date;
o.head.appendChild(o.createElement('script')).src='//www.google-analytics.com/analytics.js'}
(window,document,'g'));g('create','UA-XXXXX-X','auto');g('send','pageview');
```

__if your page must got head and you don't care about costum variable name__ - 276 characters
```
(function(b,o){b.GoogleAnalyticsObject='g';(b.g=b.g||function(){(b.g.q=b.g.q||[]).push(arguments)}).l=+new Date;
o.head.appendChild(o.createElement('script')).src='//www.google-analytics.com/analytics.js'}
(window,document));g('create','UA-XXXXX-X','auto');g('send','pageview');
```

**Comment by paulirish** [comment]
I looked at this new version and compared it to the old. The key thing here is the gzip'd size as this will "always" be served with gzip and it commonly erases many of these hand-tuning improvements.

* original GA snippet: **312** bytes, gzip
* h5bp GA snippet: **301** bytes, gzip  (Δ of 11 to orig)
* blooperz GA snippet: **283** bytes, gzip (Δ of 18 to h5bp)

(I'm gzipping the entire index.html, but probably should be. [gzip is cool](https://vimeo.com/34164210))

Anyways, we end up with a gain of 18 bytes. Due to how TCP works, the boundaries that actually end up affecting download time are at something like 14KB, 44KB, 88KB.

image
[[via](http://cdn.oreillystatic.com/en/assets/1/event/99/Prioritize%20Your%20Critical%20CSS%20and%20Images%20To%20Make%20Your%20Site%20Fast%20Presentation.pdf)]

Why? TCP-slow start, basically. 

Check the next few slides from here: https://docs.google.com/presentation/d/1qoginKKkdrRpIPGjJakt1y17iXsIBSaInaHbR4JyCIk/present?slide=id.gc03305a_0207

image


image


tl;dr: Small differences in byte size will not have a significant impact when the total filesize of the payload is within these below buckets.

* 0KB to 13KB
* 16K to 41KB
* 44KB to 99KB

The boundaries, therefore, where this sort of byte golfing does have impact are around: 14.2K, 42.7K, 99.8K.

(These calculations ignore the overhead of HTTP headers. Assuming initial_cwnd of 10 and MSS of 1460.)

File: dist/index.html [hunk 3]
```diff
@@ -29,12 +29,9 @@
 
         <!-- Google Analytics: change UA-XXXXX-X to be your site's ID. -->
         <script>
-            (function(b,o,i,l,e,r){b.GoogleAnalyticsObject=l;b[l]||(b[l]=
-            function(){(b[l].q=b[l].q||[]).push(arguments)});b[l].l=+new Date;
-            e=o.createElement(i);r=o.getElementsByTagName(i)[0];
-            e.src='//www.google-analytics.com/analytics.js';
-            r.parentNode.insertBefore(e,r)}(window,document,'script','ga'));
-            ga('create','UA-XXXXX-X','auto');ga('send','pageview');
+            (function(b,o,i,l){GoogleAnalyticsObject=l;(b[l]=b[l]||function(){(b[l].q=b[l].q||[]).push(arguments)}).l=+new Date;
```
Comment (arthurvr) [inline dist/index.html]: I think the line length was wrapped at ~80 characters by purpose.

Comment (arthurvr) [inline dist/index.html]: Why not `(function(h,5,b,p){`?

Comment (mathiasbynens) [inline dist/index.html]: Because `5` is not a valid JS identifier.

**Comment by BLooperZ** [comment]
@arthurvr - now the lines are wrapped (i wanted to make sure it still works with that whitespace...)

File: dist/index.html [hunk 4]
```diff
@@ -29,11 +29,10 @@
...
         <script>
-            (function(b,o,i,l,e,r){b.GoogleAnalyticsObject=l;b[l]||(b[l]=
-            function(){(b[l].q=b[l].q||[]).push(arguments)});b[l].l=+new Date;
-            e=o.createElement(i);r=o.getElementsByTagName(i)[0];
-            e.src='//www.google-analytics.com/analytics.js';
-            r.parentNode.insertBefore(e,r)}(window,document,'script','ga'));
+            (function(b,o,i,l){GoogleAnalyticsObject=l;(b[l]=b[l]||
+            function(){(b[l].q=b[l].q||[]).push(arguments)}).l=+new Date;
+            o.getElementsByTagName(i)[0].parentNode.appendChild(o.createElement(i))
+            .src='//www.google-analytics.com/analytics.js'}(window,document,'script','ga'));
```
Comment (mathiasbynens) [inline dist/index.html]: Let’s make this a hardcoded `https://` for the reasons outlined here: https://github.com/konklone/cdns-to-https#conclusion-cdns-should-redirect-to-https. This is a change everyone will agree on. (Perhaps I should open a separate PR for it?)

The other changes are open to bikeshedding :)

Comment (alrra) [inline dist/index.html]: @mathiasbynens Done in https://github.com/h5bp/html5-boilerplate/commit/eee759bfe175e850bbc8e4ad0682ec4fe4bd05d6.

**Comment by alrra** [comment]
@paulirish, @mathiasbynens Can any of you take care of this issue? :)

**Comment by alrra** [comment]
ping @paulirish, @mathiasbynens

**Comment by mathiasbynens** [comment]
I have a counterproposal.

```html
<script src="https://www.google-analytics.com/analytics.js" async defer></script>
```

With this snippet, modern browsers use `async`, older browsers (i.e. IE8 & IE9) use `defer`.

IE8 and IE9 lack `async` support but they have a [broken](https://github.com/h5bp/lazyweb-requests/issues/42) implementation of `defer`. AFAICT, the brokenness doesn’t apply in this scenario since no scripts depend on GA in the way jQuery UI depends on jQuery.

So why not switch to this pattern?

Once we drop support for IE8 and IE9, we could even drop the `defer` part.

The full (unminified) snippet becomes:

```html
<script>
window.GoogleAnalyticsObject = 'ga';
window.ga = function() {
  window.ga.q.push(arguments);
};
window.ga.q = [];
window.ga.l = +new Date;

ga('create', 'UA-XXXX-Y', 'auto');
ga('send', 'pageview');
</script>
<script src="https://www.google-analytics.com/analytics.js" async defer></script>
```

Minified (although byte size is not that big of a deal anymore at this point, as @paulirish demonstrated):

```html
<script>
  GoogleAnalyticsObject='ga';ga=function(){ga.q.push(arguments)};ga.q=[];ga.l=+new Date;
  ga('create','UA-XXXX-Y','auto');ga('send', 'pageview')
</script>
<script src="https://www.google-analytics.com/analytics.js" async defer></script>
```

**Comment by battaglr** [comment]
@mathiasbynens: I like the idea. :+1:

**Comment by corvannoorloos** [comment]
Would it be possible to track the user's [pageview](https://github.com/h5bp/html5-boilerplate/blob/master/dist/doc/faq.md#why-is-the-google-analytics-code-at-the-bottom-google-recommends-it-be-placed-in-the-head) if they leave before the page has loaded by moving (only) the snippet to the page header?

**Comment by sarukuku** [comment]
+1 for @mathiasbynens suggestion

**Comment by WebNinjrDesign** [comment]
Personally loving @mamathiasbynens suggestion

**Comment by d10** [comment]
:+1: to @mathiasbynens's idea!

**Comment by ramasilveyra** [comment]
I really like @mathiasbynens ga snippet.

**Comment by GloverDonovan** [comment]
Supporting **@mathiasbynens** because of its simplicity and use of async. (:

**Comment by xeniun** [comment]
So sorry because I'm not skilled with JS.
Google provide this code below as an alternative asynchronous snippet.
```html
<script async src='//www.google-analytics.com/analytics.js'></script>
<script>
window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
ga('create', 'UA-XXXX-Y', 'auto');
ga('send', 'pageview');
</script>
```
Why we can't simply add "defer" to this snippet?

**Comment by mathiasbynens** [comment]
@xeniun Are you saying Google provides that snippet? Link?

**Comment by xeniun** [comment]
Hi @mathiasbynens !
I'm relearning about new Google Universal Analytics because I need to import all my previous parameters and just today I read about this snippet here: https://developers.google.com/analytics/devguides/collection/analyticsjs/advanced in "Alternative Asynchronous Snippet" section.
I hope I have been of help.
Google is evolving and expanding rapidly their own style and help resources.

**Comment by mathiasbynens** [comment]
@xeniun Thanks for the link! I wonder why they bothered to add the IE9 comment instead of just adding `defer` to the snippet.

That snippet indicates that the `GoogleAnalyticsObject = 'ga';` part isn’t necessary, so we can remove it from our snippet:

```html
<script>
  ga=function(){ga.q.push(arguments)};ga.q=[];ga.l=+new Date;
  ga('create','UA-XXXX-Y','auto');ga('send','pageview')
</script>
<script src="https://www.google-analytics.com/analytics.js" async defer></script>
```

**Comment by xeniun** [comment]
Just out of curiosity, is the same if we reverse the sequences of the two scripts?

**Comment by mathiasbynens** [comment]
@xeniun Yep, that doesn’t really matter.

**Comment by paulirish** [comment]
oh yeah. I'm on board with using `async defer`. 

I think we should be explicit with the `ga` global and put a `window.` in front of it. Otherwise LGTM!

**Comment by mathiasbynens** [comment]
@paulirish Like this?

```html
<script>
  window.ga=function(){window.ga.q.push(arguments)};window.ga.q=[];window.ga.l=+new Date;
  window.ga('create','UA-XXXX-Y','auto');window.ga('send','pageview')
</script>
<script src="https://www.google-analytics.com/analytics.js" async defer></script>
```

Seems a bit much IMHO. I generally prefer making globals explicit in browser code by using `window.foo` too, but doing it consistently in this case adds quite a bit of unneeded overhead.

**Comment by FagnerMartinsBrack** [comment]
Considering [Paul's comment](https://github.com/h5bp/html5-boilerplate/pull/1660#issuecomment-72380964), is there any reason not to provide an unminified source in the boilerplate?

**Comment by alrra** [comment]
@mathiasbynens Can you open a pull request? Thanks!

**Comment by mathiasbynens** [comment]
@alrra Done in #1696. I haven’t added `window.` yet like @paulirish since I’m not sure where it should be added. Adding it everywhere seems overkill, and anything else seems inconsistent. https://github.com/h5bp/html5-boilerplate/pull/1660#issuecomment-89981609

**Comment by alrra** [comment]
Thanks @mathiasbynens! :purple_heart:

--

Closing in favor of #1696.

**Comment by whardson** [comment]
Hello!

 

New message, please read <http://radek.pomorze.pl/subject.php?ltmzd>

 

win@affordca.com

**Comment by whardson** [comment]
Hello!

 

New message, please read <http://mannyaniceto.com/she.php?o02p>

 

win@affordca.com

**Comment by whardson** [comment]
Hello!

 

New message, please read <http://neox.tmpl.mk.ua/offer.php?rdxhj>

 

win@affordca.com
